### DISCUSSION TITLE
Downloading go package at runtime

### DISCUSSION THREAD

**Discussion by scariabpc** [apple/pkl#788] [General]
Hi, I have a pkl module that includes

```
import "package://pkg.pkl-lang.org/pkl-go/pkl.golang@0.5.0#/go.pkl"
```

that allows us to codegen the Go code for the pkl module.

In our main.go, we have something like

```
cfg, err := pklmyapplication.LoadFromPath(mainCtx, "./pkl/values/services/myapplication.pkl")
```

We deploy this code into our IoT devices that have varying levels of internet connectivity. Occasionally, we see the following error:

```
panic: error loading config
        –– Pkl Error ––
        I/O error loading module `package://pkg.pkl-lang.org/pkl-go/pkl.golang@0.5.0#/go.pkl`.
        ZipException: zip END header not found
        
        4 | import "package://pkg.pkl-lang.org/pkl-go/pkl.golang@0.5.0#/go.pkl"
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        at ...Config#go (file:///pkl/config/services/myapplication.pkl)
        
        1 | @go.Package { name = "github.com/.../generated/pkl/go/pklmyapplication" }
             ^^^^^^^^^^
        at ....myapplication.Config (file:///pkl/config/services/myapplication.pkl)
        

goroutine 1 [running]:
main.main()
```

Can someone help me understand why we need to download the go package at runtime? Also, we have noticed that if we restart our Docker container on the device, this error goes away.

**Comment by HT154** [comment]
Because you're importing and using the package, it must be present for evaluation to complete.

If you cache the package into your image, it won't need to be downloaded in the field. You can perform this caching by adding this to your Dockerfile:
```dockerfile
RUN pkl eval package://pkg.pkl-lang.org/pkl-go/pkl.golang@0.5.0#/go.pkl
```

I would recommend doing this for any known Pkl package dependencies to avoid the need to download those packages at runtime

**Comment by scariabpc** [comment]
Awesome, thanks!
